plasmapy/diagnostics/proton_radiography.py

```diff
@@ -0,0 +1,1022 @@
+
+    def _adaptive_dt(self, Ex, Ey, Ez, Bx, By, Bz):
+        r"""
+        Calculate the appropraite dt based on a number of considerations
+        including the local grid resolution (ds) and the gyroperiod of the
+        particles in the current fields.
+        """
+        # If dt was explicitly set, skip the rest of this function
+        if self.dt.size == 1:
+            return self.dt
+
+        # Compute the timestep indicated by the grid resolution
+        ds = self.grid.grid_resolution.to(u.m).value
+        gridstep = 0.5 * (np.min(ds) / self.v0)
+
+        # If not, compute a number of possible timesteps
+        # Compute the cyclotron gyroperiod
+        Bmag = np.max(np.sqrt(Bx ** 2 + By ** 2 + Bz ** 2)).to(u.T).value
+
+        # Compute the gyroperiod
+        if Bmag == 0:
+            gyroperiod = np.inf
+        else:
+            gyroperiod = 2 * np.pi * self._m_p / (self._e * np.max(Bmag))
+
+        # TODO: introduce a minimum timestep based on electric fields too!
+
+        # Create an array of all the possible time steps we computed
+        candidates = np.array([gyroperiod / 12, gridstep])
+
+        # Enforce limits on dt
+        candidates = np.clip(candidates, self.dt[0], self.dt[1])
+
+        # dt is the min of the remaining candidates
+        return np.min(candidates)
+
+    def _coast_to_grid(self):
+        r"""
+        Coasts all particles to the timestep when the first particle should
+        be entering the grid. Doing in this in one step (rather than pushing
+        the particles through zero fields) saves computation time.
+        """
+        # Distance from the source to the nearest gridpoint
+        dist = np.min(np.linalg.norm(self.grid_arr - self.source, axis=3))
+
+        # Find the particle with the highest speed towards the grid
+        vmax = np.max(np.dot(self.v, self.src_n))
+
+        # Time for fastest possible particle to reach the grid.
+        t = dist / vmax
+
+        # Coast the particles to the advanced position
+        self.x = self.x + self.v * t
+
+    def _generate_null_distribution(self):
+        r"""
+        Calculate the distribution of particles on the detector plane in the absence
+        of any simulated fields.
+
+        These positions are used to quickly compute null radiographs, which are
+        used to determine the degree of deflection.
+        """
+        # Calculate the unit vector from the source to the detector
+        dist = np.linalg.norm(self.src_det)
+        uvec = self.src_det / dist
+
+        # Calculate the remaining distance each particle needs to travel
+        # along that unit vector
+        remaining = np.dot(self.source, uvec)
+
+        # Calculate the time remaining to reach that plane and push
+        t = (dist - remaining) / np.dot(self.v, uvec)
+
+        # Calculate the particle positions for that case
+        self.x0 = self.source + self.v * np.outer(t, np.ones(3))
```

Comment:
Since v already has the required dimensions, you can just broadcast:
```suggestion
        self.x0 = self.source + self.v * t[:, np.newaxis]
```